grpc.go

```diff
@@ -0,0 +1,698 @@
+
+	return emptyResp, nil
+}
+
+func (a *agentGRPC) WriteStdin(ctx context.Context, req *pb.WriteStreamRequest) (*pb.WriteStreamResponse, error) {
+	proc, _, err := a.pod.getRunningProcess(req.ContainerId, req.ProcessId)
+	if err != nil {
+		return &pb.WriteStreamResponse{}, err
+	}
+
+	var file *os.File
+	if proc.termMaster != nil {
+		file = proc.termMaster
+	} else {
+		file = proc.stdin
+	}
+
+	n, err := file.Write(req.Data)
+	if err != nil {
+		return &pb.WriteStreamResponse{}, err
+	}
+
+	return &pb.WriteStreamResponse{
+		Len: uint32(n),
+	}, nil
+}
+
+func (a *agentGRPC) ReadStdout(ctx context.Context, req *pb.ReadStreamRequest) (*pb.ReadStreamResponse, error) {
+	data, err := a.pod.readStdio(req.ContainerId, req.ProcessId, int(req.Len), true)
+	if err != nil {
+		return &pb.ReadStreamResponse{}, err
+	}
+
+	return &pb.ReadStreamResponse{
+		Data: data,
+	}, nil
+}
+
+func (a *agentGRPC) ReadStderr(ctx context.Context, req *pb.ReadStreamRequest) (*pb.ReadStreamResponse, error) {
+	data, err := a.pod.readStdio(req.ContainerId, req.ProcessId, int(req.Len), false)
+	if err != nil {
+		return &pb.ReadStreamResponse{}, err
+	}
+
+	return &pb.ReadStreamResponse{
+		Data: data,
+	}, nil
+}
+
+func (a *agentGRPC) CloseStdin(ctx context.Context, req *pb.CloseStdinRequest) (*gpb.Empty, error) {
+	proc, _, err := a.pod.getRunningProcess(req.ContainerId, req.ProcessId)
+	if err != nil {
+		return emptyResp, err
+	}
+
+	var file *os.File
+	if proc.termMaster != nil {
+		file = proc.termMaster
+	} else {
+		file = proc.stdin
+	}
+
+	if err := file.Close(); err != nil {
+		return emptyResp, err
+	}
+
+	return emptyResp, nil
+}
+
+func (a *agentGRPC) TtyWinResize(ctx context.Context, req *pb.TtyWinResizeRequest) (*gpb.Empty, error) {
+	proc, _, err := a.pod.getRunningProcess(req.ContainerId, req.ProcessId)
+	if err != nil {
+		return emptyResp, err
+	}
+
+	if proc.termMaster == nil {
+		return emptyResp, fmt.Errorf("Terminal is not set, impossible to resize it")
+	}
+
+	fd := int(proc.termMaster.Fd())
+
+	// Get current terminal size.
+	winSize, err := unix.IoctlGetWinsize(fd, unix.TIOCGWINSZ)
+	if err != nil {
+		return emptyResp, err
+	}
+
+	winSize.Row = uint16(req.Row)
+	winSize.Col = uint16(req.Column)
+
+	// Set new terminal size.
+	if err := unix.IoctlSetWinsize(fd, unix.TIOCSWINSZ, winSize); err != nil {
+		return emptyResp, err
+	}
+
+	return emptyResp, nil
+}
+
+func (a *agentGRPC) StartSandbox(ctx context.Context, req *pb.StartSandboxRequest) (*gpb.Empty, error) {
+	if a.pod.running == true {
+		return emptyResp, fmt.Errorf("Pod already started, impossible to start again")
+	}
+
+	a.pod.id = req.Hostname
+	a.pod.network.dns = req.Dns
+	a.pod.running = true
+
+	if err := mountShareDir(req.Storage.Source, req.Storage.MountPoint, req.Storage.Fstype); err != nil {
+		return emptyResp, err
+	}
+
+	a.pod.shareDir = req.Storage.MountPoint
+
+	if err := setupDNS(a.pod.network.dns); err != nil {
+		return emptyResp, err
+	}
+
```

Comment:
A pod pid namespace might be needed to be created. hyperstart creates one.
the pod's pid namespace's init process acts as infra container.

Or let it become an option in the StartSandboxRequest config?
@sameo @gnawux 

One more question, when infra container was requested from the k8s, will runtime ask the agent create it?

Reply:
It is already there: https://github.com/kata-containers/agent/pull/1/files#diff-727ccd4217b7f335e689be03bfb69e9bR322

Reply:
@sboeuf That one is one pid namespace per container, which is the reason docker has the pid1 problem. We need to introduce a pod level pid namespace, or at least make it configurable in `StartSandbox()`.

@laijs we have a running agent in each pod, I don't think the k8s pause container is necessary currently. But I'm not sure if k8s community will use it to do something other than zombie process recycling in future.

Reply:
https://github.com/kata-containers/agent/pull/12

Reply:
@bergwolf We got rid of the pause container in CC recently, trying to simplify things.

@sboeuf, please create a PID namespace when `sandbox_pidns` is set.

Reply:
Yes I will do that !

Reply:
@laijs @sameo I have been checking libcontainer but I am not sure it allows for this case. Just to be clear and to make sure we're all on the same page, what do we expect through this ?
My understanding is that we want our agent to create a new PID namespace and provide this namespace to every container started from this agent. How can we achieve this ?

Reply:
@sboeuf it looks like if `configs.NEWPID` is not set in the spec, libcontainer creates the new container in caller's pid namespace. Then we can create a pid namespace at `CreateSandbox()`, and enter the pid namespace before calling into libcontainer. I haven't tested if this can work though.

Reply:
@gnawux @laijs @bergwolf @sameo @mcastelino libcontainer allows us to specify a specific namespace, which is good, but we need this namespace to exist already. Creating a persistent namespace could be the solution here, but there is still a problem with that. The first process joining the namespace is going to be considered as `init` inside this namespace, meaning that if it exits, all others processes into this namespaces are going to be killed. So basically starting a container into this shared pid ns would mean that when this container terminates, all other containers are going to be terminated too, which is not the behavior we want.
The solution is to have a persistent process as the "init" process in this ns. Either we go with the agent, but this means that any other process started by the agent could be seen by containers, or we start a `pause` container (everybody loves `pause` container ;)) when we create the sandbox. This way it will join this namespace first, and we are sure it would not die.
In case of `!sharedPidNs`, then we don't do anything. This means in docker case we would have the same output from a `ps -A` while in k8s case, we would see all containers processes + this pause container process as pid 1.
WDYT ?

Reply:
Or we let the agent spawn an `init` process that handles SIGCHLD for all containers like hyperstart currently does. But I can live with a `pause` container :)

Reply:
Either way, I don't think it's a show stopper for the PR. We can add things incrementally.

Reply:
@bergwolf I said `pause` container but I was actually thinking of a simple pause binary we would start inside this specific ns. And we would kill it from `DestroySandbox()`.

Reply:
@bergwolf it does not sound too complicated to implement, and if we agree on it, I can do this directly on this PR. But if we don't agree yet, then we can postpone this to another PR.

Reply:
@bergwolf @laijs I think ideally we'd implement a separate subreaper as part of this repo, that would does only one thing: reaping all processes inside a pidns. And the agent would launch it for every pidns it has to create. Is that what the hyperstart process does ?

Reply:
@sameo Not really, the agent (not systemd nor sandbox_pidns nor any additional subreaper like containerd-shim) should be the subreaper of all the containers&processes and possible helper processes. If the agent runs as the vm's init process, like hyperstart, "agent as subreaper" was already done, in the other side when agent is not the vm's init process, some code is needed to be added.

Reply:
@sboeuf the implementation for sandbox_pidns can be postponed. Any other feature/behavior can be also be delayed to next PRs when it is complicated or under discussion. A issue might need to  be created to track the unimplemented feature/behavior.

Reply:
I have opened #20 and #21 to track that ;)